File: Valid_Anagram.js
// Time Complexity: O(n)
// Space Complexity: O(n)

function isAnagram(s, t) {
  if (s.length !== t.length) return false;
  let map = {};
  for (let i = 0; i < s.length; i++) {
    if (map[s[i]] === undefined) {
      map[s[i]] = 1;
    } else {
      map[s[i]]++;
    }
    if (map[t[i]] === undefined) {
      map[t[i]] = -1;
    } else {
      map[t[i]]--;
    }
  }
  for (let key in map) {
    if (map[key] !== 0) return false;
  }
  return true;
}

console.log(isAnagram('car', 'rat')); // true

File: valid_anagram.go
package main

// Time complexity: O(n)
// Space complexity: O(1)

func isAnagram(s string, t string) bool {
	chars := make([]int, 26)

	for _, v := range s {
		i := int(v - 'a')
		chars[i]++
	}

	for _, v := range t {
		i := int(v - 'a')
		chars[i]--
	}

	for _, v := range chars {
		if v != 0 {
			return false
		}
	}

	return true
}

// func main() {
// 	fmt.Println(isAnagram("anagram", "nagaram"))
// 	fmt.Println(isAnagram("rat", "car"))
// }
